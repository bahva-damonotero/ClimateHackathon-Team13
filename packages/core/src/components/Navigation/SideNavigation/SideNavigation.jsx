import React from 'react';
import PropTypes from 'prop-types';

import { createClassString } from '@boozallen-uip/uswds-react-utils';
import { Link } from '../..';

function renderSideNavigationItem(
  { current, href, sublist, text, className, ...props },
  key
) {
  return (
    <li className="usa-sidenav__item" key={key}>
      <Link
        className={
          current ? createClassString(['usa-current', className]) : className
        }
        href={href}
        {...props}
      >
        {text}
      </Link>

      {sublist && (
        <ul className="usa-sidenav__sublist">
          {sublist.map(renderSideNavigationItem)}
        </ul>
      )}
    </li>
  );
}

export default function SideNavigation({ items, ...props }) {
  return (
    <nav aria-label="Secondary navigation" {...props}>
      <ul className="usa-sidenav">{items.map(renderSideNavigationItem)}</ul>
    </nav>
  );
}

const SideNavigationPropType = PropTypes.shape({
  current: PropTypes.bool,
  href: PropTypes.string.isRequired,
  sublist: PropTypes.arrayOf(PropTypes.shape(this)),
  text: PropTypes.string.isRequired
});

SideNavigation.propTypes = {
  /** Array of links to show */
  items: PropTypes.arrayOf(SideNavigationPropType).isRequired
};
