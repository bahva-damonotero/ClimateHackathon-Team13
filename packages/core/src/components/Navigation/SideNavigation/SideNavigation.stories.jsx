import React from 'react';
import { MemoryRouter } from 'react-router-dom';

import SideNavigation from '.';
import { GridCol, GridRow } from '../../Grid';

export default {
  title: 'Navigation/SideNavigation',
  component: SideNavigation,
  argTypes: {
    items: {
      type: null,
      table: {
        type: {
          summary:
            '{ current: boolean, href: string, sublist: this[], text: string }[]'
        }
      }
    }
  },
  parameters: {
    controls: { hideNoControlsWarning: true }
  }
};

const Template = (args) => (
  <MemoryRouter>
    <GridRow>
      <GridCol width={4}>
        <SideNavigation {...args} />
      </GridCol>
    </GridRow>
  </MemoryRouter>
);

export const singleLevel = Template.bind({});
singleLevel.args = {
  items: [
    {
      current: true,
      href: '#',
      text: 'Current page'
    },
    {
      href: '#',
      text: 'Parent link'
    },
    {
      href: '#',
      text: 'Parent link'
    }
  ]
};

export const twoLevels = Template.bind({});
twoLevels.args = {
  items: [
    {
      href: '#',
      text: 'Parent link'
    },
    {
      current: true,
      href: '#',
      sublist: [
        {
          href: '#',
          text: 'Child link'
        },
        {
          href: '#',
          text: 'Child link'
        },
        {
          href: '#',
          text: 'Child link'
        },
        {
          href: '#',
          text: 'Child link'
        },
        {
          current: true,
          href: '#',
          text: 'Child link'
        }
      ],
      text: 'Curent page'
    },
    {
      href: '#',
      text: 'Parent link'
    }
  ]
};

export const threeLevels = Template.bind({});
threeLevels.args = {
  items: [
    {
      href: '#',
      text: 'Parent link'
    },
    {
      current: true,
      href: '#',
      sublist: [
        {
          href: '#',
          text: 'Child link'
        },
        {
          href: '#',
          sublist: [
            {
              href: '#',
              text: 'Grandchild link'
            },
            {
              href: '#',
              text: 'Grandchild link'
            },
            {
              current: true,
              href: '#',
              text: 'Grandchild link'
            }
          ],
          text: 'Child link'
        },
        {
          href: '#',
          text: 'Child link'
        },
        {
          href: '#',
          text: 'Child link'
        },
        {
          href: '#',
          text: 'Child link'
        }
      ],
      text: 'Current page'
    },
    {
      href: '#',
      text: 'Parent link'
    }
  ]
};
