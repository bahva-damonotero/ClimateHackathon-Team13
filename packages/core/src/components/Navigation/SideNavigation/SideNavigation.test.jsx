import React from 'react';
import { MemoryRouter as Router } from 'react-router-dom';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

import SideNavigation from '.';

describe('SideNavigation', () => {
  const links = [
    {
      href: '#',
      text: 'Parent link'
    },
    {
      current: true,
      href: '#',
      sublist: [
        {
          href: '#',
          text: 'Child link'
        },
        {
          href: '#',
          sublist: [
            {
              href: '#',
              text: 'Grandchild link'
            },
            {
              href: '#',
              text: 'Grandchild link'
            },
            {
              current: true,
              href: '#',
              text: 'Grandchild link'
            }
          ],
          text: 'Child link'
        },
        {
          href: '#',
          text: 'Child link'
        },
        {
          href: '#',
          text: 'Child link'
        },
        {
          href: '#',
          text: 'Child link'
        }
      ],
      text: 'Current page'
    },
    {
      href: '#',
      text: 'Parent link'
    }
  ];

  const jsx = (
    <Router>
      <SideNavigation items={links} />
    </Router>
  );

  it('should render correctly', () => {
    const { container } = render(jsx);

    // should have 3 tiers
    expect(
      container.querySelector(
        '.usa-sidenav .usa-sidenav__item .usa-sidenav__sublist .usa-sidenav__item .usa-sidenav__sublist .usa-sidenav__item'
      )
    ).toBeTruthy();
  });

  it('should match snapshot', () => {
    const tree = renderer.create(jsx).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
