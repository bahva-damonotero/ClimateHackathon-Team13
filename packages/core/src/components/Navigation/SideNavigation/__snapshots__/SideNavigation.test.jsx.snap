// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNavigation should match snapshot 1`] = `
<nav
  aria-label="Secondary navigation"
>
  <ul
    className="usa-sidenav"
  >
    <li
      className="usa-sidenav__item"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        Parent link
      </a>
    </li>
    <li
      className="usa-sidenav__item"
    >
      <a
        className="usa-current"
        href="/"
        onClick={[Function]}
      >
        Current page
      </a>
      <ul
        className="usa-sidenav__sublist"
      >
        <li
          className="usa-sidenav__item"
        >
          <a
            href="/"
            onClick={[Function]}
          >
            Child link
          </a>
        </li>
        <li
          className="usa-sidenav__item"
        >
          <a
            href="/"
            onClick={[Function]}
          >
            Child link
          </a>
          <ul
            className="usa-sidenav__sublist"
          >
            <li
              className="usa-sidenav__item"
            >
              <a
                href="/"
                onClick={[Function]}
              >
                Grandchild link
              </a>
            </li>
            <li
              className="usa-sidenav__item"
            >
              <a
                href="/"
                onClick={[Function]}
              >
                Grandchild link
              </a>
            </li>
            <li
              className="usa-sidenav__item"
            >
              <a
                className="usa-current"
                href="/"
                onClick={[Function]}
              >
                Grandchild link
              </a>
            </li>
          </ul>
        </li>
        <li
          className="usa-sidenav__item"
        >
          <a
            href="/"
            onClick={[Function]}
          >
            Child link
          </a>
        </li>
        <li
          className="usa-sidenav__item"
        >
          <a
            href="/"
            onClick={[Function]}
          >
            Child link
          </a>
        </li>
        <li
          className="usa-sidenav__item"
        >
          <a
            href="/"
            onClick={[Function]}
          >
            Child link
          </a>
        </li>
      </ul>
    </li>
    <li
      className="usa-sidenav__item"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        Parent link
      </a>
    </li>
  </ul>
</nav>
`;
